File: IAIso-v5.0/core/iaiso-php/tests/Unit/Middleware/AnthropicMiddlewareTest.php
<?php

declare(strict_types=1);

namespace IAIso\Tests\Unit\Middleware;

use IAIso\Audit\MemorySink;
use IAIso\Core\BoundedExecution;
use IAIso\Core\BoundedExecutionOptions;
use IAIso\Core\PressureConfig;
use IAIso\Core\StepInput;
use IAIso\Middleware\Anthropic\BoundedClient;
use IAIso\Middleware\Anthropic\Client;
use IAIso\Middleware\Anthropic\ContentBlock;
use IAIso\Middleware\Anthropic\Options;
use IAIso\Middleware\Anthropic\Response;
use IAIso\Middleware\EscalationRaisedException;
use PHPUnit\Framework\TestCase;

final class AnthropicMiddlewareTest extends TestCase
{
    public function testAccountsTokensAndToolCalls(): void
    {
        $sink = new MemorySink();
        $exec = BoundedExecution::start(new BoundedExecutionOptions(auditSink: $sink));

        $raw = new class implements Client {
            public function messagesCreate(array $params): Response
            {
                return new Response(
                    'claude-opus-4-7', 100, 250,
                    [
                        new ContentBlock('text'),
                        new ContentBlock('tool_use'),
                        new ContentBlock('tool_use'),
                    ],
                );
            }
        };
        $client = new BoundedClient($raw, $exec, Options::defaults());
        $client->messagesCreate([]);

        $foundStep = false;
        foreach ($sink->events() as $e) {
            if ($e->kind === 'engine.step') {
                self::assertSame(350, (int) $e->data['tokens']);
                self::assertSame(2, (int) $e->data['tool_calls']);
                $foundStep = true;
            }
        }
        self::assertTrue($foundStep, 'expected engine.step event');
        $exec->close();
    }

    public function testRaisesOnEscalationWhenOptedIn(): void
    {
        $cfg = PressureConfig::builder()
            ->escalationThreshold(0.4)->releaseThreshold(0.95)
            ->depthCoefficient(0.5)->dissipationPerStep(0.0)->build();
        $exec = BoundedExecution::start(new BoundedExecutionOptions(config: $cfg));
        $exec->recordStep(new StepInput(depth: 1));  // force escalation

        $raw = new class implements Client {
            public function messagesCreate(array $params): Response
            {
                return new Response('m', 0, 0, []);
            }
        };
        $client = new BoundedClient($raw, $exec, new Options(true));
        $this->expectException(EscalationRaisedException::class);
        $client->messagesCreate([]);
    }
}

File: IAIso-v5.0/core/iaiso-java/iaiso-middleware/src/test/java/io/iaiso/middleware/AnthropicMiddlewareTest.java
package io.iaiso.middleware;

import com.google.gson.JsonObject;
import io.iaiso.audit.MemorySink;
import io.iaiso.core.BoundedExecution;
import io.iaiso.core.BoundedExecutionOptions;
import io.iaiso.core.PressureConfig;
import io.iaiso.core.StepInput;
import io.iaiso.middleware.anthropic.AnthropicMiddleware;
import org.junit.Test;

import java.util.Arrays;

import static org.junit.Assert.*;

public class AnthropicMiddlewareTest {

    @Test
    public void accountsTokensAndToolCalls() {
        MemorySink sink = new MemorySink();
        BoundedExecution exec = BoundedExecution.start(BoundedExecutionOptions.builder()
            .auditSink(sink).build());

        AnthropicMiddleware.Client raw = params -> new AnthropicMiddleware.Response(
            "claude-opus-4-7", 100L, 250L,
            Arrays.asList(
                new AnthropicMiddleware.ContentBlock("text"),
                new AnthropicMiddleware.ContentBlock("tool_use"),
                new AnthropicMiddleware.ContentBlock("tool_use")
            ));
        AnthropicMiddleware.BoundedClient client = new AnthropicMiddleware.BoundedClient(
            raw, exec, AnthropicMiddleware.Options.defaults());
        client.messagesCreate(new JsonObject());

        boolean foundStep = false;
        for (var e : sink.events()) {
            if ("engine.step".equals(e.getKind())) {
                assertEquals(350L, ((Number) e.getData().get("tokens")).longValue());
                assertEquals(2L, ((Number) e.getData().get("tool_calls")).longValue());
                foundStep = true;
            }
        }
        assertTrue("expected engine.step event", foundStep);
        exec.close();
    }

    @Test
    public void raisesOnEscalationWhenOptedIn() {
        PressureConfig cfg = PressureConfig.builder()
            .escalationThreshold(0.4)
            .releaseThreshold(0.95)
            .depthCoefficient(0.5)
            .dissipationPerStep(0.0)
            .build();
        BoundedExecution exec = BoundedExecution.start(BoundedExecutionOptions.builder()
            .config(cfg).build());
        // Force escalation
        exec.recordStep(StepInput.builder().depth(1).build());

        AnthropicMiddleware.Client raw = params ->
            new AnthropicMiddleware.Response("model", 0, 0, java.util.Collections.emptyList());
        AnthropicMiddleware.BoundedClient client = new AnthropicMiddleware.BoundedClient(
            raw, exec, new AnthropicMiddleware.Options(true));
        try {
            client.messagesCreate(new JsonObject());
            fail("expected EscalationRaised");
        } catch (MiddlewareException.EscalationRaised expected) {}
        exec.close();
    }
}
